### PR TITLE
[Console] Convert all non-autocomplete lib files to typescript 

### PR DESCRIPTION
**Pull request by curq**
### Description

Converts ```lib/!autocomplete``` to typescript. Part of the https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3825

### Issues Resolved

Solves https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3827

### Testing the changes

No type errors present in affected files, except from imports of other non-ts files, will be fixed by other PRs from the same group of issues (https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4008, https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3826). 

### Check List

- [X] All tests pass
  - [X] `yarn test:jest`
  - [X] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [X] Update [CHANGELOG.md](./../CHANGELOG.md)
- [X] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/console/public/lib/osd/osd.ts
Comment (ananzh) [inline src/plugins/console/public/lib/osd/osd.ts:44]: nice

File: src/plugins/console/public/lib/osd/osd.ts
Comment (ananzh) [inline src/plugins/console/public/lib/osd/osd.ts:58]: I see this `ParametrizedComponentFactories` type is used in multiple places. Should we create a common folder or some type file to host it?

File: src/plugins/console/public/lib/osd/osd.ts
Comment (ananzh) [inline src/plugins/console/public/lib/osd/osd.ts:89]: `any` type seems not ideal. this is binding to ParametrizedComponentFactories context right?
should we do `getComponent(name: keyof ParametrizedComponentFactories ... )`?

File: src/plugins/console/public/lib/osd/api.ts
Comment (ananzh) [inline src/plugins/console/public/lib/osd/api.ts:38]: where is this `../autocomplete/types`? did we define it somewhere?

File: src/plugins/console/public/lib/osd/api.ts
Comment (ananzh) [inline src/plugins/console/public/lib/osd/api.ts:39]: if a type file is defined, should move `ParametrizedComponentFactories` in a single type file

File: src/plugins/console/public/lib/osd/api.ts
Comment (ananzh) [inline src/plugins/console/public/lib/osd/api.ts:31]: same lodash issue

File: src/plugins/console/public/lib/osd/api.ts
Comment (ananzh) [inline src/plugins/console/public/lib/osd/api.ts:76]: `globalRules` is defined as `Record<string, any>`. `Record<string, any>` implies that the values in `globalRules` could be of any type, so we can't directly assert that this.globalRules[term] is of type SharedComponent[], right?

File: typings/index.d.ts
Comment (ananzh) [inline typings/index.d.ts:49]: curious, why add this?

File: src/plugins/console/public/lib/curl_parsing/curl.ts
Comment (ananzh) [inline src/plugins/console/public/lib/curl_parsing/curl.ts:82]: so line could be undefined?

File: src/plugins/console/public/lib/osd/api.ts
Comment (ananzh) [inline src/plugins/console/public/lib/osd/api.ts:75]: why only throwOnMissing has `?`

**Review by ananzh — COMMENTED** [review]
I think we need to verify all the `?` usage.

File: src/plugins/console/public/lib/osd/osd.ts
Comment (curq) [reply src/plugins/console/public/lib/osd/osd.ts:212]: 1. I will update it alongside of lodash/jquery removal. I planned to make separate PRs for TS and other refactoring to make reviewing easier.
2. Yes, there is code where the functions is called with no arguments https://github.com/opensearch-project/OpenSearch-Dashboards/blob/487cd363abd91559369bb9c8fc7771175265bd0d/src/plugins/console/public/lib/osd/osd.js#L196-L198

File: src/plugins/console/public/lib/osd/osd.ts
Comment (curq) [reply src/plugins/console/public/lib/osd/osd.ts:58]: 1. You right, we should move it, I will add it to types file that I created in my other PR. I forgot to mention the order of merging, it should first be https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4008, then https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4148 and only after that this one.

File: src/plugins/console/public/lib/osd/osd.ts
Comment (curq) [reply src/plugins/console/public/lib/osd/osd.ts:89]: I agree, I will try updating it after other 2 related PRs will be merged so that I could transfer the interface to the separate file and update it there.

File: src/plugins/console/public/lib/osd/api.ts
Comment (curq) [reply src/plugins/console/public/lib/osd/api.ts:38]: Sorry, I forgot to mention that this PR should be merged after https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4148, where this file was created.

File: src/plugins/console/public/lib/osd/api.ts
Comment (curq) [reply src/plugins/console/public/lib/osd/api.ts:39]: Will do

File: src/plugins/console/public/lib/osd/api.ts
Comment (curq) [reply src/plugins/console/public/lib/osd/api.ts:76]: Yes, that's right. I referred to the type defined in file that already was in TS.
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/487cd363abd91559369bb9c8fc7771175265bd0d/src/plugins/console/server/services/spec_definitions_service.ts#L43-L45

File: typings/index.d.ts
Comment (curq) [reply typings/index.d.ts:49]: The ```curl_parsing.test``` imports txt file for the testing purposes and TS required module declaration

File: src/plugins/console/public/lib/curl_parsing/curl.ts
Comment (curq) [reply src/plugins/console/public/lib/curl_parsing/curl.ts:82]: .shift only returns undefined when lines array is empty and this condition was tested in lines above ```(lines.length === 0)```, I should change it to ! instead.

File: src/plugins/console/public/lib/osd/api.ts
Comment (curq) [reply src/plugins/console/public/lib/osd/api.ts:75]: I looked through the code and whenever this method used, the term was always present while throwOnMissing was optional.

File: src/plugins/console/public/lib/osd/api.ts
Comment (curq) [reply src/plugins/console/public/lib/osd/api.ts:83]: I think the ? is likely redundant here, I will review the ? and will remove if this is the case, thank for the catch.

File: src/plugins/console/public/lib/osd/osd.ts
Comment (curq) [reply src/plugins/console/public/lib/osd/osd.ts:183]: In all the cases where this function was used the json was always present, so I assumed it was required,

**Comment by curq** [comment]
@ananzh Thank you for the review! I responded to your comments. I forgot to initially mention the order of merging/review, it should first be https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4008, then https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4148 and only after that this one. The reason for that is that they have types on which this PR is dependent. So I think it would be better to return to this one after they are finished, because the changes there may also affect this PR.

Regarding lodash/jquey, I was planning to remove it in separate PR in order to not mix new TS logic and lodash/jquery refactoring for better review experience. If you think it is better to do it all in this PR, I'm ready to do it there, just let me know.

File: src/plugins/console/public/lib/osd/osd.ts
Comment (ananzh) [reply src/plugins/console/public/lib/osd/osd.ts:212]: 1. cool. thanks.
2. nice catch.

File: src/plugins/console/public/lib/osd/osd.ts
Comment (ananzh) [reply src/plugins/console/public/lib/osd/osd.ts:32]: sure. I am okay with that.

File: src/plugins/console/public/lib/osd/osd.ts
Comment (ananzh) [reply src/plugins/console/public/lib/osd/osd.ts:58]: I see. Then this PR is blocked by #4008 and #4148. Let's get back to this PR after them merged. @curq I will start reviewing these two.

**Comment by abbyhu2000** [comment]
The pre-req for this PR are merged: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4008 and https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4148

File: src/plugins/console/public/lib/osd/api.ts
Comment (curq) [reply src/plugins/console/public/lib/osd/api.ts]: I removed this redundant `?`. I've gone through the other usages of `?` and we do need them.

File: src/plugins/console/public/lib/osd/api.ts
Comment (curq) [reply src/plugins/console/public/lib/osd/api.ts:75]: Here is the example, where throwOnMissing is optional
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/d94d7cb683aa76878bba5120cb8130add147d7ad/src/plugins/console/public/lib/autocomplete/autocomplete.ts#L883
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/d94d7cb683aa76878bba5120cb8130add147d7ad/src/plugins/console/public/lib/autocomplete/body_completer.ts#L127

File: src/plugins/console/public/lib/osd/osd.ts
Comment (curq) [reply src/plugins/console/public/lib/osd/osd.ts]: I have updated this part and removed usage of `any`

File: src/plugins/console/public/lib/osd/osd.ts
Comment (curq) [reply src/plugins/console/public/lib/osd/osd.ts]: I moved it to `types.ts` files that holds the common types

**Comment by curq** [comment]
@abbyhu2000 Thanks for the heads up! I've addressed all the comments and updated my PR. It's now ready for review. @ananzh Could you please check it out?

**Review by ananzh — APPROVED** [review]

**Review by joshuarrrr — APPROVED** [review]
LGTM

**Pull request merged**
